### ISSUE TITLE
The <event> duration attribute specification

### ISSUE THREAD

**Issue by notator** [w3c/mnx#129]
Edit 05.06.2018: I've been continually revising this posting since it first appeared, but have not been tracking the changes because a) there have not yet been any responses, and b) previous versions are available from the "edited" menu if anyone really wants to see them.

---

The `<event> duration` attribute is a very important addition to the spec. It makes an explicit distinction between how a symbol looks and what it means, and is a very elegant way to code (Baroque) notations that existed before tuplets were invented. But I think its documentation in the spec (at the [end of §5.2.5.1](https://w3c.github.io/mnx/specification/#element-attrdef-event-duration)) could be improved.

Just above Example 23 (in §5.2.5.1), the spec says (1):
> This attribute is unlikely to be frequently encountered, but is needed to handle cases in which composers employ notated values that are not interpreted literally. Some examples follow.

Just below Example 23, the spec says (2):
> Because duration must be specified as a legal note value multiple, an enclosing tuplet element may be required in some cases to impose an appropriate time modification ratio.

These need re-phrasing and reorganising.
1. I don't think the spec should say how often this feature is likely to be used. If composers want to use an available feature, they will.
2. There is currently only one example (the "Brahms"). I think we need some more.
3. (2) could be simplified, and belongs above (1)

So I'd replace (1) by:
>Note that an enclosing tuplet element will, as usual, superimpose an appropriate time modification ratio.

>This attribute is needed to handle Baroque notations, and cases in which composers want to override the default durations of the notated symbols.
Some examples follow.

and delete (2).

All examples from the literature should be clearly identified.
I've been unable to discover where the current "Brahms" example
![image](https://user-images.githubusercontent.com/1655500/40970738-bfa56180-68bb-11e8-8664-53e07c597700.png)
originates, but it first occurs [here (in issue 79)](https://github.com/w3c/mnx/issues/79#issuecomment-373517641). So the sentence
> One case occurs in Brahms, in which a dotted half is used to specify a note that is traditionally understood as occupying the space of 11 sixteenth notes:

needs to be re-phrased and put with the "Brahms" Example:
>Derived from a Brahms Example -- quoted in [Hook](http://www.mtosmt.org/issues/mto.11.17.4/mto.11.17.4.hook.html)(?)
>Here, a dotted half-note is used to specify a note that usually occupies the space of 11 sixteenth notes 

---
**suggestions for other examples**

I think it would help if some more examples from #79 were included. Especially the two Bach examples and the [Schumann](https://github.com/w3c/mnx/issues/79#issuecomment-373510892) (which is especially interesting because it involves tuplets).

I'd start the examples with Bach:

J.S. Bach: from the Sarabande in Partita BWV 829
The top voice in the first bar of
![image](https://user-images.githubusercontent.com/1655500/40845514-37bdceb2-65b7-11e8-8f09-747b6f72b3a0.png)
would be coded like this (total 3/4 = 12/16):
```
<sequence>
  <event value="/4">
    <note pitch="F5"/>
  </event>
  <event value="/4d" duration="7/16">
    <note pitch="B4"/>
  </event>
  <event value="/8" duration="1/16">
    <note pitch="E5"/>
  </event>
</sequence>
```
--
J. S. Bach, from Die Kunst der Fuge, 1749-50
The top voice in the first bar of
![image](https://user-images.githubusercontent.com/1655500/40846401-ccdbd032-65b9-11e8-8269-9de6b91a9900.png)
would be coded like this (in 12/16 time)
```
<sequence>
  <beamed value="6/16">
    <event value="/16d" duration="1/8">
      <note pitch="F5"/>
    </event>
    <event value="/32" duration="1/16">
      <note pitch="D5"/>
    </event>
    <event value="/16d" duration="1/8">
      <note pitch="A4"/>
    </event>
    <event value="/32" duration="1/16">
      <note pitch="D5"/>
    </event>
  </beamed>
  <beamed value="6/16">
    <event value="/16" duration="3/32">
      <note pitch="F5"/>
    </event>
    <event value="/16" duration="3/32">
      <note pitch="G5"/>
    </event>
    <event value="/16" duration="3/32">
      <note pitch="F5"/>
    </event>
    <event value="/16" duration="3/32">
      <note pitch="E5"/>
    </event>
  </beamed>
</sequence>
```
Its really great, that we now have a simple, elegant solution for notating music that does not use tuplets! :-)

**Refactoring proposal**: I think `<beamed>`'s **`value`** attribute should actually be called `duration`. The [`<beamed>` definition](https://w3c.github.io/mnx/specification/#the-beamed-element) is actually rather unlear about the attribute's type (**please clarify**), but it is either a [_**note value quantity**_](https://w3c.github.io/mnx/specification/#note-value-quantity) (like a time-signature) or a simple Rational. It is definitely not a [_**note value**_](https://w3c.github.io/mnx/specification/#note-values) (like `<event>`'s **`value`** attribute).
**Correction/Suggestion**: I think a _**note value quantity**_ should always have a numerator, so [the single eighth-note example](https://w3c.github.io/mnx/specification/#note-value-quantity) should be "1/8", not "/8".

--
Schumann, Bar 7 of Bunte Blätter, Op. 99, No. 2  (quoted in [Hook](http://www.mtosmt.org/issues/mto.11.17.4/mto.11.17.4.hook.html))
![issue79schumann](https://user-images.githubusercontent.com/1655500/40969077-48d7e1f8-68b7-11e8-88d3-0a92be2911d5.png)
According to the time signature, this piece is in 2/4 time.
If the eighth-notes in the top top voice were not tuplets, they could be allocated the following  durations: 3/32,  2/32, 3/32, 3/32, 2/32, 3/32.
The durations are, however, triplets (enclosed in `<tuplet>` elements), so each inner duration has to be multiplied by 3/2, resulting in the following duration sequence:
9/64, 3/32, 9/64, 9/64, 3/32, 9/64.
Schumann could have notated such durations directly, but it would have made the graphics very complicated and difficult to read - especially at speed.
Note that the _live interpretation_ of scores relates not only to the information they contain, but also to traditions of _live performance practice_ (see [Hook](http://www.mtosmt.org/issues/mto.11.17.4/mto.11.17.4.hook.html)).
The top voice of this measure can thus be coded as follows:
```
<sequence>
  <beamed>
    <tuplet inner="3/8" outer="1/4">
      <event value="/8" duration="9/64">
        <note pitch="G4"/>
      </event>
      <event value="/8" duration="3/32">
        <note pitch="F4"/>
      </event>
      <event value="/8" duration="9/64">
        <note pitch="G4"/>
      </event>
    </tuplet>
  </beamed>
  <beamed>
    <tuplet inner="3/8" outer="1/4">
      <event value="/8" duration="9/64">
        <note pitch="A4"/>
      </event>
      <event value="/8" duration="3/32">
        <note pitch="G#4"/>
      </event>
      <event value="/8" duration="9/64">
        <note pitch="A4"/>
      </event>
    </tuplet>
  </beamed>
</sequence>
```

---
**duration overriding**

While the `<event>` duration attribute affects both the duration and the _spacing_ of events in an instantiation, a related `<performance-event>` duration attribute only affects the duration.
If both the `<event>` duration attribute is specified _and_ there is a related `<performance-event>`, the  `<performance-event>`'s duration would probably override the performed duration, without changing the spacing. This is something that should be covered in the Draft Spec.

Example 35 in [§5.2.13.1](https://w3c.github.io/mnx/specification/#the-interpret-element) shows how to describe a "swing" performance without affecting the spacing of the symbols:
```
<beamed>
  <event value="/8">
    <note pitch="C4"/>
  </event>
  <event value="/8">
    <note pitch="D4"/>
  </event>
  <interpret>
    <performance-event start="0" duration="1/6" pitch="C4"/>
    <performance-event start="1/6" duration="1/12" pitch="D4"/>
  </interpret>
</beamed>
```
Using <event> duration attributes rather than `<interpret>` and `<performance-event>`, the spacing would also be affected:
```
<beamed>
  <event value="/8" duration="1/6">
    <note pitch="C4"/>
  </event>
  <event value="/8" duration="1/12">
    <note pitch="D4"/>
  </event>
</beamed>.
```
If _both_ are present the `<event> duration` values determine the spacing, but the `<performance-event> duration` values determine the durations :
```
<beamed>
  <event value="/8" duration=1/6>
    <note pitch="C4"/>
  </event>
  <event value="/8 duration="1/12">
    <note pitch="D4"/>
  </event>
  <interpret>
    <performance-event start="0" duration="9/48"/>
    <performance-event start="1/6" duration="3/48"/>
  </interpret>
</beamed>
```

**Comment by joeberkovitz** [comment]
Good points, although we did have a longish window to review the pull request that finalized these changes -- so that would have been the best time to provide additional input, to minimize thrash.

From here I suggest the following course of action:
- Please break out a different issue for changing `value` to `duration` in the `<beamed>` element (I agree the name change makes sense for the reasons you gave). It's a very straightforward change as you say, so if you'd like to make a pull request immediately, that would probably make sense. Please keep that PR completely focused on just the one change.
- Once that's resolved, create a pull request for this issue that applies the editorial rewording you feel would be helpful along with the additional Bach example. As you say, the spec should not talk about how often a feature is used. W3C specs do not get too encyclopedic in terms of examples so we should keep this light; examples can always be supplied separately.
- Also please remove and break out any material here relating to performance event duration overriding, which is a point on which I do not at all agree (`<interpret>` information is not supposed to affect the semantic layer of the score and this is very much part of the MNX-Common architecture).  Let's all please make the best effort to create separate issues to reflect separate concerns.

A little more about the Bach Kunst example -- I'm not sure of the context, but unless the time signature for that passage is 12/8 (and perhaps it is), hidden tuplets would indeed be needed for the triplets to supply a time-modification ratio. It's not quite the case to say that `duration` removes the need for tuplets.

**Comment by notator** [comment]
Sorry, but I was away while you were discussing this.

>Please break out a different issue for changing value to duration in the <beamed> element

Will do.

>Once that's resolved, create a pull request for this issue that applies the editorial rewording you feel would be helpful along with the additional Bach example.
... please remove and break out any material here relating to performance event duration overriding.

It would be best to resolve the following points before I make the pull request:
1. The four examples: "Brahms", Bach Saraband, Bach Kunst d. F., and Schumann, illustrate progressively complex situations, so they could be put in that order. The Schumann illustrates the coding of both beams and tuplets so, if there's only space for one of them in the main text, then it ought to be the Schumann. Possibly the remaining examples should go in some kind of Appendix? How would you like to do this?

2. Do you agree that performance-event duration overrides event duration?

The Bach Kunst d.F. example: The groups of three sixteenths could, of course be coded as hidden tuplets (without bracket or number indication), but there's no reason to do so. Adding hidden triplets means that one also has to multiply the contained duration attributes by 3/2. The net result is unnecessary complication since the same durations are expressed by doing neither. The measure duration is all that needs to be considered here. The time signature might affect performance practice (beat emphasis -- the difference between 3/4 and 6/8), but I think that's beyond the scope of the current spec.

> It's not quite the case to say that duration removes the need for tuplets.

Agreed. Tuplets are required whenever one wants to have a small number and/or bracket written above the notes. But I can't see any reason for using _hidden_ tuplets.
The event duration attribute provides a powerful alternative to using tuplets, when describing durations that are not the `note value` defaults, and I can well imagine its extensive use without the use of `<tuplet>` elements. As folows:
The fact that the event duration attribute is at the event level (as opposed to tuplets, that apply to groups of notes) means that both durations _and spacing_ can be very flexibly organised.
For example, two parallel parts in a 1/4 (=250/1000) duration can be allocated the durations:
23/1000, 58/1000, 169/1000
and
31/1000, 47/1000, 23/1000, 51/1000, 98/1000
These durations would be spaced left-right by the instantiating application, as if the notes were enclosed in the appropriate (hidden) tuplets.
However, hidden tuplets are not really an option in this case, since the actual `note value` s would be prohibitively complicated!

**Comment by joeberkovitz** [comment]
- For additional examples, we do need to create a place to put them.  I suggest that as part of your PR you create an Examples appendix to the spec that contains _links only_ for brevity, which link to XML and image files in the `examples/` repo folder -- that would be a great service to everyone making future examples.

- Note value multiples do not allow denominators other than powers of 2. Hence the need for hidden tuplets in order to specify time modification ratios, even if a number is not visible. This has an impact on the examples, of course.

The chairs do not support breaking the power-of-two-denominator convention for note value multiples in CWMN. This would require all apps to make a lot of additional effort to support a much more complex and unconstrained rational-duration model, which in CWMN would only have value in these various edge cases. Hidden tuplets are needed anyway in CWMN, and they amply support the edge cases already. In fact, current notation editors appear to support these constructs today by using hidden tuplets, as a legitimate way to apply an arbitrary time ratio to a group of notes that follow the CWMN constraints that they expect.

**Comment by notator** [comment]
I don't accept that the Bach examples can be regarded as edge cases (they are standard Baroque practice), and I object to having to code music which knows no tuplets in terms of tuplets. That would be both inaccurate and inefficient.

I also don't agree that a model in which all durations are expressed as arbitrary fractions of a whole-note would be "much more complex". If applications implemented such a model (which would be quite simple), it would be very easy to implement tuplets in terms of that model.
 
In my view, no future standard should force coders to write inaccurate and/or inefficient code just because "that's the way its always been done". The argument that _current_ apps would have to be re-coded simply doesn't apply.

**Comment by joeberkovitz** [comment]
Perhaps my choice of words was unfortunate, but actually this is not a debate about what constitutes an "edge case".  The point is that with the current model, event durations in any measure or tuplet _are constrained to time intervals that can be notated as a sum of regular CWMN note values_.  Arbitrary rational durations would break that constraint.

Consider this sequence of edits to an empty 4/4 bar:
- two half notes are inserted
- the duration of the first half note is set to 3/7

At this point, what is the notated content of the bar? With the current constraint, editors can "keep the bar whole" by adding normal CWMN rests between the two notes (for example, if the first note's duration was changed to 7/16, a sixteenth rest could be added). Without that constraint, it seems far less clear what to do, yet all applications would be obliged to handle the case.

So my belief (perhaps incorrect) is that the existing constraint is a good tradeoff, and doesn't harm the need to accommodate Baroque musical practice as well as other cases that are truly "edge". Of course, I may be wrong on that point, in which case the rest of the community should chime in and support this change.

**Comment by notator** [comment]
The easiest default would be to change the duration of the second minim to 4/7,
but a rest having duration 1/14 (=1/2 - 3/7) could also be inserted if that's what the author wanted.
Conventionally, duration classes are allocated in the following ranges:
1 >= semibreve > 1/2
1/2 >= minim > 1/4
1/4 >= crotchet > 1/8
1/8 >= quaver> 1/16
1/16 >= semiquaver> 1/32
1/32 >= demisemiquaver> 1/64
etc.
So a duration of 1/14 would be a quaver by default.

**Comment by joeberkovitz** [comment]
James, it is not enough to show that some such scheme is possible (there are multiple such schemes). We need to see some backup from the rest of the CG, showing that implementors are willing to sign up for taking on this additional responsibility in every application that supports editing.

**Comment by notator** [comment]
I see no great "additional responsibility".

If arbitrary Rational duration values are allowed, it would be very easy for existing applications to write durations explicitly in the `<event>` element, rather than expecting consuming applications to keep track of (nested) `<tuplet>` attributes.
This change would be a very minor detail in MNX, compared to the rest of the spec. It could hardly be a reason for not implementing MNX export. On the contrary, I think the ability to encode arbitrary duration sequences may ultimately be the best reason _for_ implementing MNX.

Similar arguments apply for MNX import. MNX would still have `<tuplet>` elements, which work as they always did, except that there's no need to use the `<tuplet>`'s attributes to calculate the `<event>` durations.

The extended capabilities, that are enabled by allowing _arbitrary_ duration sequences to be coded, would of course require some programming, but existing applications could develop/implement those in their own time.

**Comment by clnoel** [comment]
@notator, You are overlooking the fact that basic CWMN programs rely on things that look like half notes taking up a half note of duration, unless they are notated as tuplets. It is _definitely_ an extra level of complexity to demand that all programs that can read our basic profile of CWMN be able to handle arbitrary durations, rather than easily-defined note values. These basic programs rely on having a meter that is well-defined, and have other kinds of similar constraints. Which is why we have profiles. And to answer your question about the time-signature: it depends on the profile!

On the other hand your editor (and my editor), as mature, advanced programs could handle the arbitrary durations. I'm interested in seeing if we can find a middle ground to handle the Baroque cases in a better way than putting in invisible tuplets. But to throw out the note-value concept entirely makes MNX a lot less accessible at a time when we are trying to increase interest.

**Comment by notator** [comment]
@clnoe
> You are overlooking the fact that basic CWMN programs rely on things that look like half notes taking up a half note of duration, unless they are notated as tuplets.

I don't think that's true.
How CWMN programs handle these things internally is up to them. I'm only talking about how durations should be stored in MNX.

The note-value concept is, of course, no longer quite what it was: I agree very much with the current version of the spec, that the `<event>` element should have a `duration` attribute, thus making a clear distinction between the symbol's graphics and its meaning. 

Note that in MNX, an `<event value="1/2">` (something that looks like a half-note) could very easily have a default duration of 1/2 whole note (i.e. be equivalent to `<event value="1/2" duration="1/2">` ). 

To summarize: I think that
1.  `<tuplet>` elements should only exist for _visible_ tuplets, and should just describe the tuplet's _graphics_ (the number and the bracket). In this, they would be much like `<beam>`s.
2. in MNX, all durations should be stored directly in the `<event>` elements (as arbitrary Rational values), independently of any enclosing `<tuplet>` or `<beam>` elements.
3. I see no problem with exporting or importing `<tuplets>` formatted in this way into existing CWMN applications.

**Comment by adrianholovaty** [comment]
Marking this as closed, for the sake of cleaning up our issue database. This is very outdated — it deals with the older XML version of MNX.
